### PR TITLE
Build fixes

### PR DESCRIPTION
**Pull request by exitmouse**
I gather that `main` is unstable right now, and also it seems like most of the dev work is in `cycseq`. But...
I was trying to get Tidal to build anyway, so I figured I'd attempt to send a minimal PR that makes the build commands work. With these changes, on my local machine I can run
```
cabal v2-build --allow-newer=base,template-haskell --enable-tests --enable-benchmarks all
```
and then both `cabal v2-test` and `cabal v2-bench` run successfully. Hopefully that also means the CI will work.

I'm not totally sure that the Arc benchmarks I deleted should be deleted, but I couldn't find 2.0.0 equivalents for those functions. Let me know if I should replace them.

### CODE REVIEW

File: bench/Speed/Tidal/PatternB.hs [hunk 1]
```diff
@@ -1,44 +1,15 @@
...
 
-arc1 = Arc 3 5 
+arc1 = Arc 3 5
 arc2 = Arc 4 6
 arc3 = Arc 0 1
 arc4 = Arc 1 2
 
-withQueryTimeB :: [Benchmark] 
-withQueryTimeB = 
-  [ bgroup "withQueryTime" [
```
Comment (yaxu) [inline bench/Speed/Tidal/PatternB.hs:14]: This still exists, [in Sound.Tidal.Signal.Base](https://github.com/tidalcycles/Tidal/blob/78695cd30137fabaad0164ac490ba8c0c26d6bc4/src/Sound/Tidal/Signal/Base.hs#L403-L404)

File: bench/Speed/Tidal/PatternB.hs [hunk 2]
```diff
@@ -1,44 +1,15 @@
...
-      bench "wqt whnf" $ whnf withQueryTime (*2) 
-    , bench "wqt2 whnf" $ whnf withQueryTime (+1)
-    , bench "wqt nf" $ nf withQueryTime (*2) ]
-  ]
-
-withResultArcB :: [Benchmark]
```
Comment (yaxu) [inline bench/Speed/Tidal/PatternB.hs:20]: This is now [withEventArc](https://github.com/tidalcycles/Tidal/blob/78695cd30137fabaad0164ac490ba8c0c26d6bc4/src/Sound/Tidal/Signal/Base.hs#L373-L377)

File: bench/Speed/Tidal/PatternB.hs [hunk 3]
```diff
@@ -1,44 +1,15 @@
...
-withResultArcB = 
-  [ bgroup "withResultArc" [
-      bench "wqa med" $ whnf (withResultArc arcFunc) wqaMed
-    , bench "wqa big" $ whnf (withResultArc arcFunc) wqaBig ]
-  ]
-
-withQueryArcB :: [Benchmark]
```
Comment (yaxu) [inline bench/Speed/Tidal/PatternB.hs:27]: This still exists [in Sound.Tidal.Signal.Base](https://github.com/tidalcycles/Tidal/blob/78695cd30137fabaad0164ac490ba8c0c26d6bc4/src/Sound/Tidal/Signal/Base.hs#L395-L396)

File: bench/Speed/Tidal/PatternB.hs [hunk 4]
```diff
@@ -1,44 +1,15 @@
...
-withQueryArcB = 
-  [ bgroup "withQueryArc" [
-      bench "wqa med" $ whnf (withQueryArc arcFunc) wqaMed
-    , bench "wqa big" $ whnf (withQueryArc arcFunc) wqaBig ]
-  ]
-
-subArcB :: [Benchmark]
```
Comment (yaxu) [inline bench/Speed/Tidal/PatternB.hs:34]: This is now [maybeSect in Sound.Tidal.Arc](https://github.com/tidalcycles/Tidal/blob/78695cd30137fabaad0164ac490ba8c0c26d6bc4/src/Sound/Tidal/Arc.hs#L36-L42)

**Review by yaxu — CHANGES_REQUESTED** [review]
Thanks a lot for this, it would be great to get this building again. I've identified the missing functions if you fancy fixing those.

**Comment by exitmouse** [comment]
Thanks! I'm also going to try to understand these linking failures on the ci.

**Comment by exitmouse** [comment]
Ok -- the ci check is green! The linker issue was because `dontcrash` tests `tidal-listener`, but `tidal-listener.cabal` recently added `-dynamic` in the build options, which stops ghc from building static object files. With `-dynamic-too` it should build both static and dynamic ones.
Nix and the windows build listener are still failing, but the errors are very early and don't seem to have anything to do with Tidal. Possibly bugs in the workflow files?

**Comment by exitmouse** [comment]
The nix failures seem to be from https://github.com/cachix/install-nix-action/issues/183, and the windows build ones from https://github.com/haskell/actions/issues/264. Both seem like they just need version bumps... I'll see if I can try that.

**Comment by exitmouse** [comment]
Well, now the nix build is failing because `hosc` has the wrong version in nixpkgs, and the windows build is failing because
>  when making flags consistent: warning:
>   -dynamic-too is not supported on Windows

**Comment by exitmouse** [comment]
@mitchmindtree, I'm not sure if you're still maintaining this nix flake, but it appears (to my uneducated eye) that we need a version of `hosc` that's not available in nixpkgs, and I've got no idea how to get it. The failing nix builds are the last unsuccessful checks!

**Comment by yaxu** [comment]
Thanks @exitmouse 

Yes looks like tidal is still on 1.9.2 in nixpkgs with the older hosc dependency. It seems it just grabs packages from stackage which does have the latest versions. Not sure how to get these versions bumped..

**Comment by yaxu** [comment]
@polymorphicengine do the changes to the windows build here look good to you?

**Comment by polymorphicengine** [comment]
yeah, compiling with `-dynamic-too` is right, not sure if we should disable the listener workflows at the moment though since they are all broken (they run through, but produce a binary that doesn't work properly)

**Comment by mitchmindtree** [comment]
> Yes looks like tidal is still on 1.9.2 in nixpkgs with the older hosc dependency. It seems it just grabs packages from stackage which does have the latest versions. Not sure how to get these versions bumped..

Just to clarify, the version of tidal used in the Nix flake is the copy local to this repo.

It looks like the reason the build is failing is that the latest [pinned `hosc` version in nixpkgs is 0.19.1](https://github.com/NixOS/nixpkgs/blob/d0743034d0c163484791353ebd7fbe5e5cf03a1e/pkgs/development/haskell-modules/configuration-hackage2nix/stackage.yaml#L1127), whereas Tidal 2.0 now requires `hosc >=0.20 && <0.21`.

It looks like we can update `nixpkgs` by running the `update-stackage` script while `cd`'d into the root of the `nixpkgs` repo like so:

```
./maintainers/scripts/haskell/update-stackage.sh
```

but this updates the whole set of packages to their latest versions, so might take some time to review.

There's almost certainly a way to override `hosc` with the latest version locally in our `flake.nix` without waiting for all of nixpkgs to update - there's likely some answers in here https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/haskell.section.md, i'll do some digging soon.

**Comment by mitchmindtree** [comment]
I've opened a PR against this branch that resolves the `hosc` `0.20` version issue here https://github.com/exitmouse/Tidal/pull/1.

**Comment by mitchmindtree** [comment]
Out of curiosity, should `tidal-parse` and `tidal-listener` be building with the local `main` branch version of Tidal? Or is the plan to update those later on?

E.g. it looks like [`tidal-parse` still expects Tidal 1.10](https://github.com/tidalcycles/Tidal/blob/78695cd30137fabaad0164ac490ba8c0c26d6bc4/tidal-parse/tidal-parse.cabal#L33), and after fixing this `hosc` issue which allows the tidal lib to build it looks like the `tidal-listener` nix build continues to fail with a GHC panic:

```console
tidal-listener> [1 of 5] Compiling Sound.Tidal.Listener.Config ( src/Sound/Tidal/Listener/Config.hs, dist/build/Sound/Tidal/Listener/Config.p_o, dist/build/Sound/Tidal/Listener/Config.dyn_o )
tidal-listener> <no location info>: error:
tidal-listener>     ghc: panic! (the 'impossible' happened)
tidal-listener>   (GHC version 9.2.8:
tidal-listener>         lookupGlobal
tidal-listener>   Failed to load interface for ‘GHC.Num.BigNat’
tidal-listener>   Perhaps you haven't installed the "p_dyn" libraries for package ‘ghc-bignum’?
tidal-listener>   Use -v (or `:set -v` in ghci) to see a list of the files searched for.
tidal-listener>   Call stack:
tidal-listener>       CallStack (from HasCallStack):
tidal-listener>         callStackDoc, called at compiler/GHC/Utils/Panic.hs:181:37 in ghc:GHC.Utils.Panic
tidal-listener>         pprPanic, called at compiler/GHC/Tc/Utils/Env.hs:153:32 in ghc:GHC.Tc.Utils.Env
tidal-listener> Please report this as a GHC bug:  https://www.haskell.org/ghc/reportabug
tidal-listener>
tidal-listener> [3 of 5] Compiling Sound.Tidal.Listener.Parse ( src/Sound/Tidal/Listener/Parse.hs, dist/build/Sound/Tidal/Listener/Parse.p_o, dist/build/Sound/Tidal/Listener/Parse.dyn_o )
tidal-listener> <no location info>: error:
tidal-listener>     ghc: panic! (the 'impossible' happened)
tidal-listener>   (GHC version 9.2.8:
tidal-listener>         lookupGlobal
tidal-listener>   Failed to load interface for ‘GHC.Num.BigNat’
tidal-listener>   Perhaps you haven't installed the "p_dyn" libraries for package ‘ghc-bignum’?
tidal-listener>   Use -v (or `:set -v` in ghci) to see a list of the files searched for.
tidal-listener>   Call stack:
tidal-listener>       CallStack (from HasCallStack):
tidal-listener>         callStackDoc, called at compiler/GHC/Utils/Panic.hs:181:37 in ghc:GHC.Utils.Panic
tidal-listener>         pprPanic, called at compiler/GHC/Tc/Utils/Env.hs:153:32 in ghc:GHC.Tc.Utils.Env
tidal-listener> Please report this as a GHC bug:  https://www.haskell.org/ghc/reportabug
```

**Comment by yaxu** [comment]
@mitchmindtree yep it'll take a some time to update tidal-parse. @polymorphicengine is working on fixing up tidal-listener builds, there's been some long running issues with the `hint` dependency blocking that but that might be solved soon.

**Comment by exitmouse** [comment]
Hi, just checking in. Would you like me to add a commit to this PR to remove tidal-listener from the builds, so that the checks all pass? Or do you want to leave it in until it's fixed?

**Comment by polymorphicengine** [comment]
yes, you can remove them for now! thanks!

**Comment by exitmouse** [comment]
Alright! Sorry about the scope creep on this PR, and thanks @mitchmindtree for the nix changes!

**Comment by yaxu** [comment]
Thanks all, is this ready to merge @exitmouse ?

**Comment by exitmouse** [comment]
All good on my end. 👍

**Comment by yaxu** [comment]
Thanks again!

**Pull request merged**